File: Backend/app/models/user_model.py
from bson import ObjectId # type: ignore
from app.db import get_db
from typing import Optional 
from app.enums.roles import Role

db = get_db()

class User:
    def __init__(self, data: dict):
        if not isinstance(data, dict): data = {} 
        self.id: Optional[str] = str(data.get("_id")) if data.get("_id") else None
        self.role: str = data.get("role", Role.STUDENT.value)
        self.username: str = data.get("username", "")
        self.email: str = data.get("email", "")
        self.password:Optional[str] = data.get("password")

    def to_dict(self, include_password: bool = False) -> dict:
        user_dict  = {
            "role": self.role,
            "id": self.id,
            "username": self.username,
            "email": self.email
        }
        if include_password and self.password:
            user_dict["password"] = self.password
        return user_dict

File: Frontend/models/userModel.ts
import type { User } from "~/types/models/User";
import { Role } from "~/types/models/User";

export class UserModel implements User {
  _id?: string;
  role: Role;
  username: string;
  email?: string;
  password?: string;

  constructor(data: Partial<User> = {}) {
    this._id = data._id;
    this.role = data.role ?? Role.student;
    this.username = data.username?.trim() ?? "";
    this.email = data.email?.trim();
    this.password = data.password?.trim();
  }

  toDict(includePassword = false): Record<string, any> {
    const result: Record<string, any> = {
      role: this.role,
      username: this.username,
    };

    if (this._id) result._id = this._id;
    if (this.email) result.email = this.email;
    if (includePassword && this.password) result.password = this.password;

    return result;
  }
}
